Add KafkaAdmin class

llamahunter opened the pull request:
Requires cluster version > 0.10.0.0, and uses new wire protocol
classes to do many things via broker connection that previously
needed to be done directly in zookeeper.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dpkp/kafka-python/1540)
<!-- Reviewable:end -->

llamahunter commented:
Not really a python pro, so please provide feedback on conventions, etc.  Also, not really sure how to do extensive unit testing.  The existing tests seem very integration-centric, and don't actually run successfully for me on my machine.

dpkp commented:
This is great work! Thanks so much for submitting. My only top-level question is how do you think we should handle api version selection in a multi-version cluster? I.e., during a cluster upgrade. In the current non-admin case we provide a way for users to pass in an explicit "broker version". A more robust approach might be to always check the version of the broker we are sending the request to, but that gets significantly more complicated. Finally, maybe the last option is to advertise in the documentation that it is not safe to use while the cluster is in mixed-version / rolling upgrade state. What do you think?

llamahunter commented:
Well, I'm using the ApiVersions data structure returned by the broker to select the max version that both the library and the broker supports for each individual API call.  I think hard coding that "0.11.0.2 means XYZ parameters on operations ABC are supported" is error prone, especially when the brokers *already* provide us with exactly what versions of each API call they support.

 I think that if someone is doing an admin operation on a mixed version cluster, they should be sure that they don't use any of the optional parameters that aren't supported by all brokers in the cluster. Tho, I think that most (all?) admin calls go to the cluster controller, so it really only matters what that broker version is at.

llamahunter commented:
Hmm.  I was fetching the ApiVersions map from whatever happened to be the 'least loaded node', but that's not necessarily the controller.  All the admin calls go to the controller, so really need to check what ApiVersions it supports.  Moved the check_version() call there.

llamahunter commented:
So, looking for some feedback on whether I should augment the API with return types for the calls.  Right now, it's sort of just delivering back whatever Kafka wire protocol object.  But, it looks like that's only got python attributes for the top level, and subobjects are just tuples or something, making them difficult to use.  Thoughts?  Is there a way to make the subobjects returned by the Schema parser have attributes?  Should I unmarshal into typed python objects?

dpkp commented on kafka/admin/config_resource.py:
```diff
@@ -0,0 +1,31 @@
+from __future__ import absolute_import
+
+from enum import IntEnum
```
We have a pretty strict rule against mandatory external dependencies. I know this is standard in py3, but it isn't in py2. Can we avoid using it? If not, we should vendor the py27 version (see kafka/vendor)

llamahunter replied:
ugh... people are still using python 2?  Isn't it EOL?  Hrmph...  lemme look at it.

jeffwidman replied:
Any progress @llamahunter ?

And yes, I know of multiple companies who heavily rely on this library and still have large codebases in python 2. Most are planning to migrate to python 3, but still in the early stages. I'd guess it'll be another 2-3 years before I'd be okay with saying that we as a core infra library are going to stop supporting python 2 in our releases. I look forward to that day, it's just not quite here yet.

sbreidbach-edr replied:
FWIW, I'd love to see this be part of the project as well.

jeffwidman commented:
Hey @llamahunter I'd really like to land this. I think it'd pair well with #1196 as of a 2.0 release... 

Do you think you'll have time to look at the `enum` part? 

Personally, I'd probably just vendor it, since it's in the stdlib. That way whenever we do eventually drop python 2 support it'll be an easy change, plus in the meantime we can make additional changes throughout the code to use `enum`.

llamahunter commented:
Sorry about the delay. Will look into this today. Still would like feedback about the return format. Should there be structured data structures rather than the loosely typed property bags? Is there some better way of using the Schema class to make the properties easier to access?

jeffwidman commented:
I just added `enum34` in #1604--use it as a fallback: 

    # enum in stdlib as of py3.4
    try:
        import enum  # pylint: disable=import-error
    except ImportError:
        # vendored backport module
        from kafka.vendor import enum34 as enum

Re: the schema stuff--good question that I haven't had time to look into yet... @dpkp / @tvoinarovskyi your thoughts?

jeffwidman commented on tox.ini:
```diff
@@ -21,6 +21,7 @@ deps =
     crc32c
     py26: unittest2
     decorator
+    py27: enum34
```
since this is vendored, no need to include this

jeffwidman commented:
@llamahunter let's just ship as is for now to land the existing changes and then leave figuring out a cleaner design for a later PR... I would not be surprised if we get a few bug reports about this, so I'd rather get it live in the hands of users on master sooner rather than later, even if the interface changes slightly in a later release.

Besides, with the 2.0 release impending once I get #1196, it will be an convenient place to do a breaking change in this interface if needed. 

So if you can rebase and use the vendored `enum34` when `enum` is not available, than I think that's all that is needed to get this merged.

llamahunter commented:
Ah, sorry if I conflicted.  I rebased and removed the tox.ini change.  Had some weird local problems running the tests, tho.  Hope it works on travis.

jeffwidman commented:
Also, I think the test failures are my fault from https://github.com/dpkp/kafka-python/pull/1605 so I will get those fixed in a separate PR and then merge this...

Github was down, so travis wasn't running the tests correctly and I didn't realize it til now

llamahunter commented:
Ok... so, what needs doing here?  Do I need to add
```
# enum in stdlib as of py3.4
try:
    import enum  # pylint: disable=import-error
except ImportError:
    # vendored backport module
    from kafka.vendor import enum34 as enum
```
for my use of IntEnum type?

jeffwidman commented:
No, I already got that change in there. I was waiting until I had a chance to fix the `pylint` issue which was breaking tests... that is handled in https://github.com/dpkp/kafka-python/pull/1609 / 0d766cf5c16cbc68d42732d044f8faa0af318b0f, so I just rebased and am just waiting for tests to pass and then I will merge this.

jeffwidman commented:
Tests took forever to get green due to repeated flakiness upstream of travis--failed downloads etc.
